src/Plutarch/Test/QuickCheck/Instances.hs

```diff
@@ -0,0 +1,974 @@
+        | otherwise = variant 0
+
+isRight = flip pmatchT $ \case
+    PRight _ -> pconstant True
+    _ -> pconstant False
+pright = flip pmatchT $ \case
+    PRight a -> a
+    _ -> ptraceError "asked for PRight when it is PLeft"
+pleft = flip pmatchT $ \case
+    PLeft a -> a
+    _ -> ptraceError "asked for PLeft when it is PRight"
+
+-- | @since 2.0.0
+instance (PArbitrary a, PArbitrary b) => PArbitrary (PEither a b) where
+    parbitrary = do
+        (TestableTerm x) <- parbitrary
+        (TestableTerm y) <- parbitrary
+        elements [TestableTerm $ pcon $ PRight x, TestableTerm $ pcon $ PLeft y]
+
+    pshrink x
+        | pliftT $ isRight x =
+            [ pconT $ PRight a
+            | (TestableTerm a) <- shrink (pright x)
+            ]
+        | otherwise =
+            [ pconT $ PLeft a
+            | (TestableTerm a) <- shrink (pleft x)
+            ]
+
+instance (PCoArbitrary a, PCoArbitrary b) => PCoArbitrary (PEither a b) where
+    pcoarbitrary x
+        | pliftT $ isRight x = variant 0 . pcoarbitrary (pright x)
+        | otherwise = variant 1 . pcoarbitrary (pleft x)
+
+{- | Unfortunately, it is impossible to create @PBuiltinPair@ at the
+     Plutarch level without getting into manipulating raw Plutus
+     data. Instead, it can only be created from haskell level value
+     and constanted.
+
+     This limitation limites this generator to only accepting Plutarch
+     types that have @PLift@ and @Arbitrary (PLifted a)@.
+
+ @since 2.0.0
+-}
+instance
+    ( PLift a
+    , PLift b
+    , Arbitrary (PLifted a, PLifted b)
+    ) =>
+    PArbitrary (PBuiltinPair a b)
+    where
+    parbitrary = pconstantT <$> arbitrary
+
+    pshrink = fmap pconstantT . shrink . pliftT
+
+instance
+    ( PLift a
+    , PLift b
+    , CoArbitrary (PLifted a, PLifted b)
+    ) =>
+    PCoArbitrary (PBuiltinPair a b)
+    where
+    pcoarbitrary = coarbitrary . pliftT
+
+-- | @since 2.0.0
+instance
+    {-# OVERLAPPING #-}
+    ( PArbitrary a
+    , PArbitrary b
+    , PIsData a
+    , PIsData b
+    ) =>
+    PArbitrary (PBuiltinPair (PAsData a) (PAsData b))
+    where
+    parbitrary = do
+        (TestableTerm x) <- parbitrary
+        return $ TestableTerm $ pfromData $ pbuiltinPairFromTuple (pdata x)
+
+    pshrink = fmap fromTuple . shrink . toTuple
+      where
+        toTuple = liftTestable (pfromData . ptupleFromBuiltin . pdata)
+        fromTuple = liftTestable (pfromData . pbuiltinPairFromTuple . pdata)
+
+instance
+    {-# OVERLAPPING #-}
+    ( PCoArbitrary a
+    , PCoArbitrary b
+    , PIsData a
+    , PIsData b
+    ) =>
+    PCoArbitrary (PBuiltinPair (PAsData a) (PAsData b))
+    where
+    pcoarbitrary (liftTestable ptupleFromBuiltin . pdataT -> t) = pcoarbitrary t
+
+ppFstT :: TestableTerm (PPair a b) -> TestableTerm a
+ppFstT = flip pmatchT $ \(PPair a _) -> a
+ppSndT :: TestableTerm (PPair a b) -> TestableTerm b
+ppSndT = flip pmatchT $ \(PPair _ a) -> a
```

Comment:
Spacing. Also, if we have a bunch of such functionality we don't export, putting it in a ``-- Helpers`` section would be good.